## Dataset 1: Food Balance (Food)

In [2]:
import pandas as pd

fdbalance_df = pd.read_csv("FAOSTAT_data_commodity balances_food.csv")
print(fdbalance_df.head())

  Domain Code                 Domain  Area Code (M49)   Area  Element Code  \
0         FBS  Food Balances (2010-)              404  Kenya           511   
1         FBS  Food Balances (2010-)              404  Kenya           511   
2         FBS  Food Balances (2010-)              404  Kenya           511   
3         FBS  Food Balances (2010-)              404  Kenya           511   
4         FBS  Food Balances (2010-)              404  Kenya           511   

                         Element Item Code (FBS)        Item  Year Code  Year  \
0  Total Population - Both sexes           S2501  Population       2010  2010   
1  Total Population - Both sexes           S2501  Population       2011  2011   
2  Total Population - Both sexes           S2501  Population       2012  2012   
3  Total Population - Both sexes           S2501  Population       2013  2013   
4  Total Population - Both sexes           S2501  Population       2014  2014   

      Unit     Value Flag                   

This is your strongest feature:

👉 National_Stock_Deficit_Ratio
* (Consumption - Production) / Production

In [3]:
fdbalance_df.columns


Index(['Domain Code', 'Domain', 'Area Code (M49)', 'Area', 'Element Code',
       'Element', 'Item Code (FBS)', 'Item', 'Year Code', 'Year', 'Unit',
       'Value', 'Flag', 'Flag Description', 'Note'],
      dtype='str')

In [4]:
fdbalance_df.info()


<class 'pandas.DataFrame'>
RangeIndex: 20167 entries, 0 to 20166
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Domain Code       20167 non-null  str    
 1   Domain            20167 non-null  str    
 2   Area Code (M49)   20167 non-null  int64  
 3   Area              20167 non-null  str    
 4   Element Code      20167 non-null  int64  
 5   Element           20167 non-null  str    
 6   Item Code (FBS)   20167 non-null  str    
 7   Item              20167 non-null  str    
 8   Year Code         20167 non-null  int64  
 9   Year              20167 non-null  int64  
 10  Unit              20167 non-null  str    
 11  Value             20167 non-null  float64
 12  Flag              20167 non-null  str    
 13  Flag Description  20167 non-null  str    
 14  Note              0 non-null      float64
dtypes: float64(2), int64(4), str(9)
memory usage: 2.3 MB


In [5]:
fdbalance_df['Element'].unique()
# fdbalance_df['Item'].unique()

<StringArray>
[         'Total Population - Both sexes',
                             'Production',
                        'Import quantity',
                        'Stock Variation',
                        'Export quantity',
               'Domestic supply quantity',
                                   'Feed',
                                   'Seed',
                                 'Losses',
                             'Processing',
                  'Other uses (non-food)',
                    'Tourist consumption',
                              'Residuals',
                                   'Food',
    'Food supply quantity (kg/capita/yr)',
          'Food supply (kcal/capita/day)',
                     'Food supply (kcal)',
 'Protein supply quantity (g/capita/day)',
            'Protein supply quantity (t)',
     'Fat supply quantity (g/capita/day)',
                'Fat supply quantity (t)']
Length: 21, dtype: str

In [6]:
fdbalance_df['Item'].unique()

<StringArray>
[                     'Population',              'Wheat and products',
               'Rice and products',             'Barley and products',
              'Maize and products',                'Rye and products',
                            'Oats',             'Millet and products',
            'Sorghum and products',                  'Cereals, other',
            'Cassava and products',           'Potatoes and products',
                  'Sweet potatoes',                            'Yams',
                    'Roots, Other',                      'Sugar cane',
                      'Sugar beet',           'Sugar non-centrifugal',
          'Sugar (Raw Equivalent)',               'Sweeteners, Other',
                           'Honey',                           'Beans',
                            'Peas',      'Pulses, Other and products',
               'Nuts and products',                       'Soyabeans',
                      'Groundnuts',                  'Sunflower

Transform this dataset

In [7]:
print("shape:", fdbalance_df.shape)
print("\ncolumns:")
print(fdbalance_df.columns.tolist())

print("\nunique units:")
print(fdbalance_df["Unit"].value_counts(dropna=False))

print("\nunique elements:")
print(fdbalance_df["Element"].value_counts(dropna=False))

shape: (20167, 15)

columns:
['Domain Code', 'Domain', 'Area Code (M49)', 'Area', 'Element Code', 'Element', 'Item Code (FBS)', 'Item', 'Year Code', 'Year', 'Unit', 'Value', 'Flag', 'Flag Description', 'Note']

unique units:
Unit
1000 t          11548
g/cap/d          2448
t                2448
kg/cap           1261
kcal/cap/d       1224
million Kcal     1224
1000 No            14
Name: count, dtype: int64

unique elements:
Element
Domestic supply quantity                  1343
Export quantity                           1272
Food                                      1261
Food supply quantity (kg/capita/yr)       1261
Import quantity                           1257
Food supply (kcal/capita/day)             1224
Food supply (kcal)                        1224
Protein supply quantity (g/capita/day)    1224
Protein supply quantity (t)               1224
Fat supply quantity (g/capita/day)        1224
Fat supply quantity (t)                   1224
Residuals                                 1189


Step 2: Filter only the useful rows

From your output, we can clearly see:

❌ You have nutrition + per capita + totals mixed in
✔️ We ONLY want physical quantities for supply-demand
✅ Keep ONLY unit = "1000 t"

This removes:

g/cap/d
kcal/cap/d
kg/cap
etc.

👉 These are NOT useful for your model

✅ Keep ONLY key economic elements

In [8]:
elements_to_keep = [
    "Production",
    "Import quantity",
    "Export quantity",
    "Domestic supply quantity",
    "Food",
    "Feed",
    "Seed",
    "Processing",
    "Losses",
    "Stock Variation",
    "Other uses (non-food)"
]

In [9]:
# Filter by unit
fdbalance_clean = fdbalance_df[fdbalance_df["Unit"] == "1000 t"]

# Filter by elements
elements_to_keep = [
    "Production",
    "Import quantity",
    "Export quantity",
    "Domestic supply quantity",
    "Food",
    "Feed",
    "Seed",
    "Processing",
    "Losses",
    "Stock Variation",
    "Other uses (non-food)"
]

fdbalance_clean = fdbalance_clean[
    fdbalance_clean["Element"].isin(elements_to_keep)
]

print("new shape:", fdbalance_clean.shape)
print("\nremaining elements:")
print(fdbalance_clean["Element"].value_counts())

new shape: (9289, 15)

remaining elements:
Element
Domestic supply quantity    1343
Export quantity             1272
Food                        1261
Import quantity             1257
Stock Variation             1147
Production                  1112
Losses                       667
Processing                   393
Feed                         349
Seed                         263
Other uses (non-food)        225
Name: count, dtype: int64


🎯 What this step does

After this:

You remove all noise
You keep only actual supply-demand flows
You prepare the data for pivoting (next step)

Step 3: Pivot (this is the transformation step)

Now we convert from:

long format (Element as rows)
➡️ to
wide format (Element as columns)

In [10]:
fdbalance_pivot = fdbalance_clean.pivot_table(
    index=["Area", "Item", "Year"],
    columns="Element",
    values="Value",
    aggfunc="sum"
).reset_index()

# remove column name grouping
fdbalance_pivot.columns.name = None

print("shape:", fdbalance_pivot.shape)
print("\ncolumns:")
print(fdbalance_pivot.columns.tolist())

shape: (1343, 14)

columns:
['Area', 'Item', 'Year', 'Domestic supply quantity', 'Export quantity', 'Feed', 'Food', 'Import quantity', 'Losses', 'Other uses (non-food)', 'Processing', 'Production', 'Seed', 'Stock Variation']


🎯 What should happen

You should now see columns like:

Area
Item
Year
Production
Import quantity
Export quantity
Food
Feed
Processing
Losses
etc.

👉 Each row = one crop in one year

That is the moment where the dataset becomes actually usable for modelling and analysis. It is clean,
structured and usable for ML and economic reasoning

Step 4: Final cleaning

Right now:

column names have spaces → not ideal for coding

In [11]:
fdbalance_final = fdbalance_pivot.copy()

fdbalance_final.columns = [
    col.lower().replace(" ", "_").replace("(", "").replace(")", "")
    for col in fdbalance_final.columns
]

print(fdbalance_final.columns.tolist())

['area', 'item', 'year', 'domestic_supply_quantity', 'export_quantity', 'feed', 'food', 'import_quantity', 'losses', 'other_uses_non-food', 'processing', 'production', 'seed', 'stock_variation']


Step 5: Create your core features (this is the magic)

Now we create actual intelligence variables for AgriBrain.

In [12]:
# Supply
fdbalance_final["supply"] = (
    fdbalance_final["production"].fillna(0) +
    fdbalance_final["import_quantity"].fillna(0)
)

# Demand
fdbalance_final["demand"] = (
    fdbalance_final["food"].fillna(0) +
    fdbalance_final["feed"].fillna(0) +
    fdbalance_final["processing"].fillna(0) +
    fdbalance_final["export_quantity"].fillna(0) +
    fdbalance_final["losses"].fillna(0) +
    fdbalance_final["other_uses_non-food"].fillna(0)
)

# Market balance (VERY IMPORTANT)
fdbalance_final["balance"] = fdbalance_final["supply"] - fdbalance_final["demand"]

print(fdbalance_final[["item", "year", "supply", "demand", "balance"]].head())

                item  year  supply  demand  balance
0  Alcohol, Non-Food  2010    18.0    19.0     -1.0
1  Alcohol, Non-Food  2011    22.0    21.0      1.0
2  Alcohol, Non-Food  2012    22.0    21.0      1.0
3  Alcohol, Non-Food  2013    17.0    17.0      0.0
4  Alcohol, Non-Food  2014    36.0    28.0      8.0


🎯 What you just created

This is huge:

supply → how much exists
demand → how much is used
balance →
positive → oversupply (price drop risk)
negative → shortage (opportunity crop)

👉 This is literally your Cobweb detection engine

Step 6 (optional but powerful)

You can later classify:

In [13]:
# fdbalance_final["market_signal"] = fdbalance_final["balance"].apply(
#    lambda x: "oversupply" if x > 0 else "shortage")

Big picture (important)

With just this dataset you can already:

identify saturated crops
identify opportunity crops
feed signals into your LLM

That means this dataset is now properly cleaned and feature-engineered.

But other_uses_non-food kept the hyphen, which can be annoying later in Python formulas. It is better to replace - as well.

In [14]:
# make one final clean copy
fdbalance_final = fdbalance_pivot.copy()

# cleaner column names
fdbalance_final.columns = [
    col.lower()
       .replace(" ", "_")
       .replace("(", "")
       .replace(")", "")
       .replace("-", "_")
    for col in fdbalance_final.columns
]

# create core features
fdbalance_final["supply"] = (
    fdbalance_final["production"].fillna(0) +
    fdbalance_final["import_quantity"].fillna(0)
)

fdbalance_final["demand"] = (
    fdbalance_final["food"].fillna(0) +
    fdbalance_final["feed"].fillna(0) +
    fdbalance_final["processing"].fillna(0) +
    fdbalance_final["export_quantity"].fillna(0) +
    fdbalance_final["losses"].fillna(0) +
    fdbalance_final["other_uses_non_food"].fillna(0)
)

fdbalance_final["balance"] = fdbalance_final["supply"] - fdbalance_final["demand"]

fdbalance_final["market_signal"] = fdbalance_final["balance"].apply(
    lambda x: "oversupply" if x > 0 else ("shortage" if x < 0 else "balanced")
)

# preview
print(fdbalance_final[["item", "year", "supply", "demand", "balance", "market_signal"]].head())

# save
fdbalance_final.to_csv("fdbalance_final.csv", index=False)

print("\nSaved as: fdbalance_final.csv")
print("Final shape:", fdbalance_final.shape)

                item  year  supply  demand  balance market_signal
0  Alcohol, Non-Food  2010    18.0    19.0     -1.0      shortage
1  Alcohol, Non-Food  2011    22.0    21.0      1.0    oversupply
2  Alcohol, Non-Food  2012    22.0    21.0      1.0    oversupply
3  Alcohol, Non-Food  2013    17.0    17.0      0.0      balanced
4  Alcohol, Non-Food  2014    36.0    28.0      8.0    oversupply


PermissionError: [Errno 13] Permission denied: 'fdbalance_final.csv'

In [ ]:
print(fdbalance_final.columns.tolist())
print(fdbalance_final.head())

['area', 'item', 'year', 'domestic_supply_quantity', 'export_quantity', 'feed', 'food', 'import_quantity', 'losses', 'other_uses_non_food', 'processing', 'production', 'seed', 'stock_variation', 'supply', 'demand', 'balance', 'market_signal']
    area               item  year  domestic_supply_quantity  export_quantity  \
0  Kenya  Alcohol, Non-Food  2010                       5.0             14.0   
1  Kenya  Alcohol, Non-Food  2011                       1.0             20.0   
2  Kenya  Alcohol, Non-Food  2012                       1.0             20.0   
3  Kenya  Alcohol, Non-Food  2013                       0.0             17.0   
4  Kenya  Alcohol, Non-Food  2014                      12.0             16.0   

   feed  food  import_quantity  losses  other_uses_non_food  processing  \
0   NaN   NaN              5.0     NaN                  5.0         NaN   
1   NaN   NaN              6.0     NaN                  1.0         NaN   
2   NaN   NaN              6.0     NaN             

What we did right
Structure is clean: area, item, year → perfect indexing
Features created correctly: supply, demand, balance
Market signal works → this is your economic logic layer
Shape (1343 rows) makes sense → ~98 items × ~13–14 years

👉 This dataset is now model-ready + LLM-ready

⚠️ One important improvement (very quick)

Right now you have items like:

Alcohol, Non-Food
some non-agricultural / irrelevant categories

👉 These will confuse your model and recommendations

One important improvement (very quick)

Right now you have items like:

Alcohol, Non-Food
some non-agricultural / irrelevant categories

👉 These will confuse your model and recommendations

In [ ]:
# remove clearly irrelevant items
exclude_keywords = [
    "alcohol",
    "beverages",
    "stimulants",
    "miscellaneous",
    "non-food"
]

fdbalance_final = fdbalance_final[
    ~fdbalance_final["item"].str.lower().str.contains("|".join(exclude_keywords))
]

print("new shape after filtering:", fdbalance_final.shape)

new shape after filtering: (1287, 18)


🎯 Why this matters

You want AgriBrain to recommend:

maize
beans
capsicum
rice

NOT:

alcohol
processed categories

👉 This keeps your recommendations real-world usable

In [ ]:
# can also rank crops by opportunity:
# Higher = more shortage = better crop to recommend
fdbalance_final["opportunity_score"] = -fdbalance_final["balance"]

Final verdict on this dataset--we went from: 1343 → 1287 rows: only ~56 rows removed

That is small (~4%)

* ✔️ Fully cleaned
* ✔️ Properly engineered
* ✔️ Directly supports your objective
* ✔️ Ready to integrate with:

** EcoCrop (feasibility)
** WFP (prices)
** ML model (yield)

Why those rows needed to go

Items like:

Alcohol, Non-Food
Beverages
Miscellaneous

👉 are aggregated or processed categories, not crops a farmer can plant.

If you keep them:

your model may recommend nonsense
your LLM may say “plant alcohol” 😅

In [ ]:
print(fdbalance_final["item"].unique()[:20])

['Apples and products' 'Aquatic Animals, Others' 'Aquatic Plants'
 'Bananas' 'Barley and products' 'Beans' 'Beer' 'Bovine Meat'
 'Butter, Ghee' 'Cassava and products' 'Cephalopods' 'Cereals, other'
 'Citrus, Other' 'Cloves' 'Cocoa Beans and products' 'Coconut Oil'
 'Coconuts - Incl Copra' 'Coffee and products' 'Cottonseed'
 'Cottonseed Oil']


## Dataset ---Producer prices

In [ ]:
import pandas as pd

ppi_df = pd.read_csv("FAOSTAT_data_producer prices.csv")
print(ppi_df.head())

  Domain Code           Domain  Area Code (M49)   Area  Element Code  \
0          PP  Producer Prices              404  Kenya          5539   
1          PP  Producer Prices              404  Kenya          5539   
2          PP  Producer Prices              404  Kenya          5539   
3          PP  Producer Prices              404  Kenya          5539   
4          PP  Producer Prices              404  Kenya          5539   

                                  Element Item Code (CPC)  \
0  Producer Price Index (2014-2016 = 100)           01654   
1  Producer Price Index (2014-2016 = 100)           01654   
2  Producer Price Index (2014-2016 = 100)           01654   
3  Producer Price Index (2014-2016 = 100)           01654   
4  Producer Price Index (2014-2016 = 100)           01654   

                                                Item  Year Code  Year  \
0  Anise, badian, coriander, cumin, caraway, fenn...       1991  1991   
1  Anise, badian, coriander, cumin, caraway, fenn...  

In [ ]:
print(ppi_df.columns)

Index(['Domain Code', 'Domain', 'Area Code (M49)', 'Area', 'Element Code',
       'Element', 'Item Code (CPC)', 'Item', 'Year Code', 'Year',
       'Months Code', 'Months', 'Unit', 'Value', 'Flag', 'Flag Description'],
      dtype='object')


In [ ]:
print(ppi_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8125 entries, 0 to 8124
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Domain Code       8125 non-null   object 
 1   Domain            8125 non-null   object 
 2   Area Code (M49)   8125 non-null   int64  
 3   Area              8125 non-null   object 
 4   Element Code      8125 non-null   int64  
 5   Element           8125 non-null   object 
 6   Item Code (CPC)   8125 non-null   object 
 7   Item              8125 non-null   object 
 8   Year Code         8125 non-null   int64  
 9   Year              8125 non-null   int64  
 10  Months Code       8125 non-null   int64  
 11  Months            8125 non-null   object 
 12  Unit              4485 non-null   object 
 13  Value             8125 non-null   float64
 14  Flag              8125 non-null   object 
 15  Flag Description  8125 non-null   object 
dtypes: float64(1), int64(5), object(10)
memory

In [ ]:
ppi_df['Element'].unique()


array(['Producer Price Index (2014-2016 = 100)',
       'Producer Price (LCU/tonne)', 'Producer Price (SLC/tonne)',
       'Producer Price (USD/tonne)'], dtype=object)

In [ ]:
ppi_df['Item'].unique()

Step 1: Inspect properly


In [ ]:
print("shape:", ppi_df.shape)

print("\nunique elements:")
print(ppi_df["Element"].value_counts())

print("\nunique units:")
print(ppi_df["Unit"].value_counts(dropna=False))

print("\nsample items:")
print(ppi_df["Item"].unique()[:20])


shape: (8125, 16)

unique elements:
Producer Price Index (2014-2016 = 100)    3640
Producer Price (LCU/tonne)                1917
Producer Price (USD/tonne)                1284
Producer Price (SLC/tonne)                1284
Name: Element, dtype: int64

unique units:
NaN    3640
LCU    1917
SLC    1284
USD    1284
Name: Unit, dtype: int64

sample items:
['Anise, badian, coriander, cumin, caraway, fennel and juniper berries, raw'
 'Apples' 'Apricots' 'Areca nuts' 'Asparagus' 'Avocados' 'Bananas'
 'Barley' 'Beans, dry' 'Beeswax' 'Broad beans and horse beans, green'
 'Cabbages' 'Carrots and turnips' 'Cashew nuts, in shell' 'Cassava, fresh'
 'Castor oil seeds' 'Cauliflowers and broccoli' 'Chick peas, dry'
 'Chillies and peppers, dry (Capsicum spp., Pimenta spp.), raw'
 'Chillies and peppers, green (Capsicum spp. and Pimenta spp.)']


Final decision (important)

* We will NOT use the index
* We will use:
* ✔️ Producer Price (LCU/tonne) → BEST option

Why LCU is best
* LCU = Local Currency Unit (KES for Kenya)
* Directly usable for:
* ✔️profitability
* ✔️comparisons
* No need for exchange rate conversion and this is exactly what our model needs

Step 2: Filter

In [ ]:
# Keep only LCU prices
ppi_clean = ppi_df[
    ppi_df["Element"] == "Producer Price (LCU/tonne)"
]

print("shape:", ppi_clean.shape)
print("\ncheck units:")
print(ppi_clean["Unit"].value_counts())

shape: (1917, 16)

check units:
LCU    1917
Name: Unit, dtype: int64


Expected
* Only LCU values remain
* Much smaller dataset

Step 3: Filter to annual data

In [ ]:
ppi_clean = ppi_clean[ppi_clean["Months"] == "Annual value"]

print("shape after annual filter:", ppi_clean.shape)
print("\nmonths check:")
print(ppi_clean["Months"].value_counts())

shape after annual filter: (1284, 16)

months check:
Annual value    1284
Name: Months, dtype: int64


What we are checking
* Only “Annual value” remains
* Shape should reduce slightly (or stay similar)*--1284 rows is sensibl 

Step 4: keep only the columns we need

In [ ]:
ppi_clean = ppi_clean[[
    "Area",
    "Item",
    "Year",
    "Value"
]]

print(ppi_clean.head())
print(ppi_clean.columns.tolist())

      Area       Item  Year     Value
35   Kenya     Apples  2011   26655.0
49   Kenya     Apples  2023   97186.7
50   Kenya     Apples  2024  120101.4
162  Kenya  Asparagus  2011   10091.8
163  Kenya  Asparagus  2023  220000.0
['Area', 'Item', 'Year', 'Value']


Step 4 check
* Clean structure → ✔️
* Columns correct → ✔️
* Values look realistic → ✔️

* ✔️ we now have actual crop prices in Kenya (very important dataset)

Step 5: Rename column

In [ ]:
ppi_final = ppi_clean.rename(columns={
    "Value": "price_lcu_per_tonne"
})

print(ppi_final.head())
print(ppi_final.columns.tolist())

      Area       Item  Year  price_lcu_per_tonne
35   Kenya     Apples  2011              26655.0
49   Kenya     Apples  2023              97186.7
50   Kenya     Apples  2024             120101.4
162  Kenya  Asparagus  2011              10091.8
163  Kenya  Asparagus  2023             220000.0
['Area', 'Item', 'Year', 'price_lcu_per_tonne']


What this does
* Makes the meaning of the column explicit
* Avoids confusion later when merging datasets

Final step: Save

In [ ]:
ppi_final.to_csv("producer_prices_final.csv", index=False)

print("Saved as producer_prices_final.csv")
print("Final shape:", ppi_final.shape)

Saved as producer_prices_final.csv
Final shape: (1284, 4)


Quick final check on 
* how many crops we have
* time coverage

This dataset gives:
* actual crop prices (KES/tonne)
* per item + year
* 👉 This is the profitability backbone

In [ ]:
print(ppi_final["Item"].nunique())
print(ppi_final["Year"].min(), "-", ppi_final["Year"].max())

87
1991 - 2024


If you want this dataset to match your other cleaned files, then do one more tidy step before saving:

In [ ]:
ppi_final.columns = [
    col.lower().replace(" ", "_")
    for col in ppi_final.columns
]

print(ppi_final.columns.tolist())
print(ppi_final["item"].nunique())
print(ppi_final["year"].min(), "-", ppi_final["year"].max())

['area', 'item', 'year', 'price_lcu_per_tonne']
87
1991 - 2024


In [ ]:
ppi_final.head()

,area,item,year,price_lcu_per_tonne
35,Kenya,Apples,2011,26655.0
49,Kenya,Apples,2023,97186.7
50,Kenya,Apples,2024,120101.4
162,Kenya,Asparagus,2011,10091.8
163,Kenya,Asparagus,2023,220000.0


Save

In [ ]:
ppi_final.to_csv("producer_prices_final.csv", index=False)
print("Saved as producer_prices_final.csv")
print("Final shape:", ppi_final.shape)

Saved as producer_prices_final.csv
Final shape: (1284, 4)


# Dataset 4: fertilizer nutrient

🎯 First: What is the role of fertilizer data in AgriBrain?

It helps you:

understand input intensity
explain yield differences
possibly improve yield prediction model

👉 Not for recommendation directly, but for:

ML features
economic reasoning (cost/productivity)

Fertilizer by Nutrient ✅

Examples:

Nitrogen (N)
Phosphorus (P)
Potassium (K)

👉 This is MUCH better because:

directly linked to soil fertility
aligns with:
your soil API (nitrogen, phosphorus)
your ML model inputs

In [ ]:
import pandas as pd

fertntrnt_df = pd.read_csv("FAOSTAT_fertiliser by nutrient.csv")
print(fertntrnt_df.head())

  Domain Code                   Domain  Area Code (M49)   Area  Element Code  \
0         RFN  Fertilizers by Nutrient              404  Kenya          5610   
1         RFN  Fertilizers by Nutrient              404  Kenya          5157   
2         RFN  Fertilizers by Nutrient              404  Kenya          5159   
3         RFN  Fertilizers by Nutrient              404  Kenya          5172   
4         RFN  Fertilizers by Nutrient              404  Kenya          5173   

                                    Element  Item Code  \
0                           Import quantity       3102   
1                          Agricultural Use       3102   
2                  Use per area of cropland       3102   
3                            Use per capita       3102   
4  Use per value of agricultural production       3102   

                          Item  Year Code  Year    Unit    Value Flag  \
0  Nutrient nitrogen N (total)       1961  1961       t  3000.00    I   
1  Nutrient nitrogen N (

In [ ]:
fertntrnt_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1087 entries, 0 to 1086
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Domain Code       1087 non-null   object 
 1   Domain            1087 non-null   object 
 2   Area Code (M49)   1087 non-null   int64  
 3   Area              1087 non-null   object 
 4   Element Code      1087 non-null   int64  
 5   Element           1087 non-null   object 
 6   Item Code         1087 non-null   int64  
 7   Item              1087 non-null   object 
 8   Year Code         1087 non-null   int64  
 9   Year              1087 non-null   int64  
 10  Unit              1087 non-null   object 
 11  Value             1087 non-null   float64
 12  Flag              1087 non-null   object 
 13  Flag Description  1087 non-null   object 
 14  Note              261 non-null    object 
dtypes: float64(1), int64(5), object(9)
memory usage: 127.5+ KB


In [ ]:
fertntrnt_df['Element'].unique()
# fertntrnt_df['Item'].unique()

array(['Import quantity', 'Agricultural Use', 'Use per area of cropland',
       'Use per capita', 'Use per value of agricultural production',
       'Export quantity', 'Production'], dtype=object)

In [ ]:
fertntrnt_df['Item'].unique()

array(['Nutrient nitrogen N (total)', 'Nutrient phosphate P2O5 (total)',
       'Nutrient potash K2O (total)'], dtype=object)

Step 1 -inspect

In [ ]:
print("shape:", fertntrnt_df.shape)

print("\nunique units:")
print(fertntrnt_df["Unit"].value_counts())

print("\nunique elements:")
print(fertntrnt_df["Element"].value_counts())

print("\nsample items:")
print(fertntrnt_df["Item"].unique())

shape: (1087, 15)

unique units:
t         520
kg/ha     189
kg/cap    189
g/Int$    189
Name: Unit, dtype: int64

unique elements:
Import quantity                             189
Agricultural Use                            189
Use per value of agricultural production    189
Use per area of cropland                    189
Use per capita                              189
Export quantity                             102
Production                                   40
Name: Element, dtype: int64

sample items:
['Nutrient nitrogen N (total)' 'Nutrient phosphate P2O5 (total)'
 'Nutrient potash K2O (total)']


Step 2 

We will keep ONLY:

* Unit = "t"--will keep
* Elements:
*"Agricultural Use" ✅ (main one)
*"Import quantity" (optional)

Why this matters (quick clarity)

This dataset is not like food balance.

* 👉 It is NOT crop-level
( 👉 It is country-level input intensity

So later it will be used as:

ML feature (e.g. higher nitrogen → higher yield potential)

Step 2: Filter by unit

In [ ]:
# Filter by unit
fertntrnt_clean = fertntrnt_df[fertntrnt_df["Unit"] == "t"]

# Keep only key elements
elements_to_keep = [
    "Agricultural Use",
    "Import quantity"
]

fertntrnt_clean = fertntrnt_clean[
    fertntrnt_clean["Element"].isin(elements_to_keep)
]

print("new shape:", fertntrnt_clean.shape)
print("\nremaining elements:")
print(fertntrnt_clean["Element"].value_counts())

new shape: (378, 15)

remaining elements:
Import quantity     189
Agricultural Use    189
Name: Element, dtype: int64


🎯 What this step does

You now isolate:

* total fertilizer usage (by nutrient)
* remove all per-capita / per-area noise

⚠️ Important understanding

Each row now represents:
* Kenya
*Year
* Nutrient (N, P, K)
* Value (tonnes used)

👉 This is perfect for ML features

Step 3: Pivot

Now we convert nutrients into columns.

In [ ]:
fertntrnt_pivot = fertntrnt_clean.pivot_table(
    index=["Area", "Year"],
    columns="Item",
    values="Value",
    aggfunc="sum"
).reset_index()

# remove column grouping
fertntrnt_pivot.columns.name = None

print("shape:", fertntrnt_pivot.shape)
print("\ncolumns:")
print(fertntrnt_pivot.columns.tolist())

shape: (63, 5)

columns:
['Area', 'Year', 'Nutrient nitrogen N (total)', 'Nutrient phosphate P2O5 (total)', 'Nutrient potash K2O (total)']


🎯 What you should see

Columns like:

Area
Year
Nutrient nitrogen N (total)
Nutrient phosphate P2O5 (total)
Nutrient potash K2O (total)

👉 Each row = Kenya in a given year

⚠️ Important note

We did NOT include Element in pivot.

👉 Why?
Because:

“Agricultural Use” is what we care about
Import quantity is secondary (we may add later separately if needed)

Step 4: Final cleaning

Let us make this dataset clean and usable.

In [ ]:
fertntrnt_final = fertntrnt_pivot.copy()

# clean column names
fertntrnt_final.columns = [
    col.lower()
       .replace(" ", "_")
       .replace("(", "")
       .replace(")", "")
    for col in fertntrnt_final.columns
]

# rename nutrients (very important for readability)
fertntrnt_final = fertntrnt_final.rename(columns={
    "nutrient_nitrogen_n_total": "nitrogen",
    "nutrient_phosphate_p2o5_total": "phosphorus",
    "nutrient_potash_k2o_total": "potassium"
})

print(fertntrnt_final.columns.tolist())
print(fertntrnt_final.head())

['area', 'year', 'nitrogen', 'phosphorus', 'potassium']
    area  year  nitrogen  phosphorus  potassium
0  Kenya  1961    6000.0     14000.0     2200.0
1  Kenya  1962    7000.0     15000.0     2200.0
2  Kenya  1963   14648.0     15070.0     2332.0
3  Kenya  1964   20664.0     16147.0     1724.0
4  Kenya  1965   26000.0     20028.0     1692.0


Step 5: Save

In [ ]:
fertntrnt_final.to_csv("fertilizer_nutrient_final.csv", index=False)

print("Saved as fertilizer_nutrient_final.csv")
print("Final shape:", fertntrnt_final.shape)

Saved as fertilizer_nutrient_final.csv
Final shape: (63, 5)


🧠 Why this is powerful

This will be used for:

* ML model (yield prediction)
* explaining productivity differences
* enriching LLM reasoning (e.g. “low nitrogen use → lower yields”)
* 🚨 One small insight (important)

This dataset is:

* country-level (Kenya)
* not crop-level

👉 So later you will:

* merge on year
* not on item

✅ Final verdict

* ✔️ This dataset is very useful
* ✔️ Cleaned correctly
* ✔️ Ready for integration and This dataset is fully ready for SQL and modelling

Just ensure types are correct:

In [ ]:
fertntrnt_final["year"] = fertntrnt_final["year"].astype(int)

How this will be used later (important)

You will join it like this:

Food balance → (area, item, year)
Fertilizer → (area, year)

👉 So fertilizer becomes:

a year-level feature applied to all crops

In [ ]:
fertntrnt_final.to_csv("fertilizer_nutrient_final.csv", index=False)

In [ ]:
import os
print(os.listdir())

['eco.ipynb', 'EcoCrop.csv', 'FAOSTAT_data_commodity balances_food.csv', 'FAOSTAT_data_commodity balances_non_food.csv', 'FAOSTAT_data_cpi.csv', 'FAOSTAT_data_exchange rates.csv', 'FAOSTAT_data_macro indicators.csv', 'FAOSTAT_data_producer prices.csv', 'FAOSTAT_fertiliser by nutrient.csv', 'FAOSTAT_fertiliser by product.csv', 'FAOSTAT_pesticides use.csv', 'fdbalance_final.csv', 'fertilizer_nutrient_final.csv', 'location.py', 'notebook.ipynb', 'wfp_markets.csv']


## Dataset 5: pesticide

In [ ]:
import pandas as pd
pest_df = pd.read_csv("FAOSTAT_pesticides use.csv")
print(pest_df.head())

  Domain Code          Domain  Area Code (M49)   Area  Element Code  \
0          RP  Pesticides Use              404  Kenya          5157   
1          RP  Pesticides Use              404  Kenya          5159   
2          RP  Pesticides Use              404  Kenya          5172   
3          RP  Pesticides Use              404  Kenya          5173   
4          RP  Pesticides Use              404  Kenya          5157   

                                    Element  Item Code                Item  \
0                          Agricultural Use       1357  Pesticides (total)   
1                  Use per area of cropland       1357  Pesticides (total)   
2                            Use per capita       1357  Pesticides (total)   
3  Use per value of agricultural production       1357  Pesticides (total)   
4                          Agricultural Use       1357  Pesticides (total)   

   Year Code  Year    Unit   Value Flag Flag Description  \
0       1990  1990       t  444.00    E  Est

In [ ]:
pest_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 272 entries, 0 to 271
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Domain Code       272 non-null    object 
 1   Domain            272 non-null    object 
 2   Area Code (M49)   272 non-null    int64  
 3   Area              272 non-null    object 
 4   Element Code      272 non-null    int64  
 5   Element           272 non-null    object 
 6   Item Code         272 non-null    int64  
 7   Item              272 non-null    object 
 8   Year Code         272 non-null    int64  
 9   Year              272 non-null    int64  
 10  Unit              272 non-null    object 
 11  Value             272 non-null    float64
 12  Flag              272 non-null    object 
 13  Flag Description  272 non-null    object 
 14  Note              272 non-null    object 
dtypes: float64(1), int64(5), object(9)
memory usage: 32.0+ KB


In [ ]:
pest_df['Element'].unique()
# pest_df['Item'].unique()

array(['Agricultural Use', 'Use per area of cropland', 'Use per capita',
       'Use per value of agricultural production'], dtype=object)

In [ ]:
pest_df['Item'].unique()

array(['Pesticides (total)', 'Insecticides', 'Herbicides',
       'Fungicides and Bactericides', 'Rodenticides'], dtype=object)

First quick interpretation (so you understand fast)

This dataset is:

* 👉 Total pesticide use (Kenya, by year)
* 👉 NOT broken down by crop

So similar to fertilizer:
* it is a country-level input variable

What we already see
From your preview:
Units:
* t → ✔️ good
* kg/ha, kg/cap, g/Int$ → ❌ noise

Elements:
* Agricultural Use → ✔️ main one
* others → ❌ not needed

Items:
* Only: "Pesticides (total)"

👉 This makes things very easy

Step 1: Full inspection

In [ ]:
print("shape:", pest_df.shape)

print("\nunique units:")
print(pest_df["Unit"].value_counts())

print("\nunique elements:")
print(pest_df["Element"].value_counts())

print("\nsample items:")
print(pest_df["Item"].unique())

shape: (272, 15)

unique units:
t         170
kg/ha      34
kg/cap     34
g/Int$     34
Name: Unit, dtype: int64

unique elements:
Agricultural Use                            170
Use per area of cropland                     34
Use per value of agricultural production     34
Use per capita                               34
Name: Element, dtype: int64

sample items:
['Pesticides (total)' 'Insecticides' 'Herbicides'
 'Fungicides and Bactericides' 'Rodenticides']


🎯 Expected decision (you can already see it)

We will KEEP:
* Unit → "t"
* Element → "Agricultural Use"
Why this dataset is useful
It gives:
* intensity of pesticide use over time
* proxy for: farming intensity, pest pressure and input usage

👉 Good feature for:
* ML model
* LLM reasoning

Important difference from previous datasets
* Food balance → crop-level
* Fertilizer → nutrient-level
* Pesticides → single total value per year

👉 So final table will be:
| area | year | pesticides_total |


Decision (important nuance here)

We will NOT just take “total”. but we will KEEP breakdown:
* Pesticides (total)
* Insecticides
* Herbicides
* Fungicides and Bactericides

❌ Drop:
* Rodenticides (small + less relevant for crops)

Why this is better; Instead of:
* one number (too basic)

You now get:
* type of pest pressure
* richer ML features
* better LLM reasoning (e.g. “fungicide use high → disease pressure”)

Step 2: Filter

In [ ]:
# Filter by unit
pest_clean = pest_df[pest_df["Unit"] == "t"]

# Keep only Agricultural Use
pest_clean = pest_clean[pest_clean["Element"] == "Agricultural Use"]

# Keep relevant pesticide types
items_to_keep = [
    "Pesticides (total)",
    "Insecticides",
    "Herbicides",
    "Fungicides and Bactericides"
]

pest_clean = pest_clean[pest_clean["Item"].isin(items_to_keep)]

print("new shape:", pest_clean.shape)
print("\nremaining items:")
print(pest_clean["Item"].value_counts())

new shape: (136, 15)

remaining items:
Pesticides (total)             34
Herbicides                     34
Fungicides and Bactericides    34
Insecticides                   34
Name: Item, dtype: int64


What this does is that we now isolate:
* real usage (not ratios)
* meaningful pesticide categories

Step 3: Pivot
* Now we convert pesticide types into columns.

In [ ]:
pest_pivot = pest_clean.pivot_table(
    index=["Area", "Year"],
    columns="Item",
    values="Value",
    aggfunc="sum"
).reset_index()

# remove column grouping
pest_pivot.columns.name = None

print("shape:", pest_pivot.shape)
print("\ncolumns:")
print(pest_pivot.columns.tolist())

shape: (34, 6)

columns:
['Area', 'Year', 'Fungicides and Bactericides', 'Herbicides', 'Insecticides', 'Pesticides (total)']


Expected output is Columns like:
* Area
* Year
* Pesticides (total)
* Insecticides
* Herbicides
* Fungicides and Bactericides

Each row = Kenya per year

Step 4: Final cleaning

In [ ]:
pest_final = pest_pivot.copy()

# clean column names
pest_final.columns = [
    col.lower()
       .replace(" ", "_")
       .replace("(", "")
       .replace(")", "")
    for col in pest_final.columns
]

# rename for simplicity
pest_final = pest_final.rename(columns={
    "fungicides_and_bactericides": "fungicides",
    "pesticides_total": "pesticides_total"
})

print(pest_final.columns.tolist())
print(pest_final.head())

['area', 'year', 'fungicides', 'herbicides', 'insecticides', 'pesticides_total']
    area  year  fungicides  herbicides  insecticides  pesticides_total
0  Kenya  1990       214.0        95.0         134.0             444.0
1  Kenya  1991       481.0       215.0         302.0             998.0
2  Kenya  1992       653.0       290.0         410.0            1354.0
3  Kenya  1993       697.0       301.0         434.0            1432.0
4  Kenya  1994      1790.0       726.0        1112.0            3628.0


Step 5: Save

In [ ]:
pest_final.to_csv("pesticides_final.csv", index=False)

print("Saved as pesticides_final.csv")
print("Final shape:", pest_final.shape)

Saved as pesticides_final.csv
Final shape: (34, 6)


What we now have

| area | year | fungicides | herbicides | insecticides | pesticides_total |

Why this is strong is because we now have:
* fertilizer → soil productivity inputs
* pesticides → pest/disease pressure inputs

These two together:
* significantly improve your ML model
* strengthen your LLM explanations

One small note is that
* Rows = 34 → fewer years than fertilizer and that is fine, just means:
* later merge will have some missing years
* we can handle that with joins/fill later

Why 34 rows is fine
This dataset is:
* year-level only (Kenya)
* not crop-level

So:
* 34 rows = ~34 years of data
* that is completely normal

Compare with your other datasets
* Food balance → 1343 rows (many crops × many years)
* Fertilizer → 63 rows (years only)
* Pesticides → 34 rows (years only)

Different datasets = different levels of detail
What matters is NOT row count
What matters is:
* ✔ clean structure
* ✔ correct variables
* ✔ usable for merging

And yours checks all three.
* How it will be used later (this is key)
* When you merge:
* food_balance (item + year)    + fertilizer (year) + pesticides (year)

pesticides data will be:
* joined on year
* repeated across all crops for that year


Big reassurance
* You are not building a dataset with equal rows everywhere.
You are building:
* a multi-layer system with
* crop-level layer
* country-level layer
* time-series layer

## Dataset 2: CPI (Consumer Price Index)

This is a monthly food CPI index, not crop-specific prices. That means it will be useful later to adjust prices for inflation, but only after we reduce it to a clean annual series.

This CPI dataset will not connect to SQL the same way as the crop-level tables.
How it connects in SQL:
* CPI is a year-level macro table, just like fertilizer and pesticides. So you would store it in its own table, something like:

cpi_table
---------
* area
* year
* food_cpi

Then:
* food_balance_table joins on (area, year, item)
* producer_prices_table joins on (area, year, item)
* fertilizer_table joins on (area, year)
* pesticides_table joins on (area, year)
* cpi_table joins on (area, year)

So CPI gets joined by year, and its value is repeated across all crops for that year

How it is used in the model

CPI is not a direct agronomic feature like soil, fertilizer, or pesticides. It is mainly an economic adjustment feature.

It helps in two ways:

* Deflating nominal prices: if producer prices rise over time, CPI helps you tell whether that is a real increase or just inflation.
* Improving market reasoning: your LLM or analysis layer can distinguish “crop prices are rising because of real scarcity” from “prices are rising because overall food prices are inflating.” That supports the project’s market-analysis and anti-cobweb logic.

How it would use it practically--For the ML side, CPI is usually:
* optional for the yield model
* useful for a price / profitability model
* useful for the LLM decision layer

So:
* Yield prediction model → probably do not prioritize CPI
* Profitability / market opportunity logic → yes, include CPI
* SQL warehouse / feature store → yes, keep it

Simple example
* If maize price in 2024 is much higher than in 2012:
* without CPI, you may think maize became far more profitable
* with CPI, you may find most of that increase is just inflation

So CPI helps create a better variable like:
* real_price = nominal_price / (food_cpi / 100)

In [ ]:
import pandas as pd

cpi_df = pd.read_csv("FAOSTAT_data_cpi.csv")
print(cpi_df.head())

  Domain Code                  Domain  Area Code (M49)   Area  Year Code  \
0          CP  Consumer Price Indices              404  Kenya       2000   
1          CP  Consumer Price Indices              404  Kenya       2000   
2          CP  Consumer Price Indices              404  Kenya       2000   
3          CP  Consumer Price Indices              404  Kenya       2000   
4          CP  Consumer Price Indices              404  Kenya       2000   

   Year  Item Code                                        Item  Months Code  \
0  2000      23013  Consumer Prices, Food Indices (2015 = 100)         7001   
1  2000      23013  Consumer Prices, Food Indices (2015 = 100)         7002   
2  2000      23013  Consumer Prices, Food Indices (2015 = 100)         7003   
3  2000      23013  Consumer Prices, Food Indices (2015 = 100)         7004   
4  2000      23013  Consumer Prices, Food Indices (2015 = 100)         7005   

     Months  Element Code Element Unit      Value Flag  \
0   Januar

In [ ]:
cpi_df.columns

Index(['Domain Code', 'Domain', 'Area Code (M49)', 'Area', 'Year Code', 'Year',
       'Item Code', 'Item', 'Months Code', 'Months', 'Element Code', 'Element',
       'Unit', 'Value', 'Flag', 'Flag Description', 'Note'],
      dtype='object')

In [ ]:
cpi_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 913 entries, 0 to 912
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Domain Code       913 non-null    object 
 1   Domain            913 non-null    object 
 2   Area Code (M49)   913 non-null    int64  
 3   Area              913 non-null    object 
 4   Year Code         913 non-null    int64  
 5   Year              913 non-null    int64  
 6   Item Code         913 non-null    int64  
 7   Item              913 non-null    object 
 8   Months Code       913 non-null    int64  
 9   Months            913 non-null    object 
 10  Element Code      913 non-null    int64  
 11  Element           913 non-null    object 
 12  Unit              296 non-null    object 
 13  Value             913 non-null    float64
 14  Flag              913 non-null    object 
 15  Flag Description  913 non-null    object 
 16  Note              617 non-null    object 
dt

### # To get Real Price Feature

Real Price = Producer Price / CPI

In [ ]:
cpi_df['Element'].unique()

array(['Value'], dtype=object)

In [ ]:
cpi_df['Item'].unique()

array(['Consumer Prices, Food Indices (2015 = 100)',
       'Consumer Prices, General Indices (2015 = 100)',
       'Food price inflation'], dtype=object)

Step 1: inspect it properly

In [ ]:
print("shape:", cpi_df.shape)

print("\nunique items:")
print(cpi_df["Item"].value_counts(dropna=False))

print("\nunique elements:")
print(cpi_df["Element"].value_counts(dropna=False))

print("\nunique months:")
print(cpi_df["Months"].value_counts(dropna=False).head(15))

print("\nunique units:")
print(cpi_df["Unit"].value_counts(dropna=False))

shape: (913, 17)

unique items:
Consumer Prices, General Indices (2015 = 100)    309
Consumer Prices, Food Indices (2015 = 100)       308
Food price inflation                             296
Name: Item, dtype: int64

unique elements:
Value    913
Name: Element, dtype: int64

unique months:
June         77
August       77
July         77
February     77
April        77
May          77
March        77
January      77
September    75
December     74
October      74
November     74
Name: Months, dtype: int64

unique units:
NaN    617
%      296
Name: Unit, dtype: int64


From your output:
* Items:
* ✔ Food CPI index → keep
* ✔ General CPI → optional
* ❌ Food price inflation (%) → drop
* Units:
* ✔ NaN → index values (this is what we want)
* ✔ % → inflation → ❌ not needed
* Months:
* ✔ Monthly data → we must convert to annual

✅ Decision (very clear)
* We will KEEP: Consumer Prices, Food Indices (2015 = 100)
* We will:drop inflation (%) and convert monthly → annual average

Step 2: Filter dataset

In [ ]:
# keep only food CPI index
cpi_clean = cpi_df[
    cpi_df["Item"] == "Consumer Prices, Food Indices (2015 = 100)"
]

# keep only index values (Unit = NaN)
cpi_clean = cpi_clean[cpi_clean["Unit"].isna()]

print("shape:", cpi_clean.shape)
print("\nitems check:")
print(cpi_clean["Item"].value_counts())

shape: (308, 17)

items check:
Consumer Prices, Food Indices (2015 = 100)    308
Name: Item, dtype: int64


🎯 What this does is that we now isolate:
* only food-related inflation signal
* remove % inflation noise

Step 3: Convert monthly → annual average

In [ ]:
cpi_yearly = cpi_clean.groupby(
    ["Area", "Year"]
)["Value"].mean().reset_index()

print("shape:", cpi_yearly.shape)
print(cpi_yearly.head())

shape: (26, 3)
    Area  Year      Value
0  Kenya  2000  21.378676
1  Kenya  2001  23.005705
2  Kenya  2002  23.606976
3  Kenya  2003  26.719484
4  Kenya  2004  30.979831


🎯 What this does
* Takes all months in a year
* Computes average CPI for that year
* 👉 This is what we need for:
* inflation adjustment
* joining with other datasets

Expected result
* Rows ≈ number of years-- 26 rows → ~26 years → ✔️ correct
* Columns:
* Area
* Year
* Value (annual CPI)

Step 4: Final cleaning

In [ ]:
cpi_final = cpi_yearly.copy()

# rename column
cpi_final = cpi_final.rename(columns={
    "Value": "food_cpi"
})

# clean column names
cpi_final.columns = [
    col.lower().replace(" ", "_")
    for col in cpi_final.columns
]

print(cpi_final.head())
print(cpi_final.columns.tolist())

    area  year   food_cpi
0  Kenya  2000  21.378676
1  Kenya  2001  23.005705
2  Kenya  2002  23.606976
3  Kenya  2003  26.719484
4  Kenya  2004  30.979831
['area', 'year', 'food_cpi']


🎯 What you now have

| area | year | food_cpi |

👉 Clean, simple, and ready for merging

Step 5: Save

In [ ]:
cpi_final.to_csv("cpi_final.csv", index=False)

print("Saved as cpi_final.csv")
print("Final shape:", cpi_final.shape)

Saved as cpi_final.csv
Final shape: (26, 3)


🧠 Now let me connect it back (as promised, clearly)

This CPI will be used to:

1. Adjust prices (very important later)
* You can compute: real price
* 👉 real_price = price_lcu_per_tonne / (food_cpi / 100)
* 👉 This gives inflation-adjusted prices

2. Improve model reasoning
* 👉 Your system can distinguish:
* “prices are high because demand > supply”
* vs
* “prices are high because inflation increased”

3. SQL usage
* 👉Table:
* cpi_table
* area
* year
* food_cpi

* 👉 Join:

* JOIN ON (area, year)

Big picture--we have built:
* market structure (food balance)
* prices (producer prices)
* inflation (CPI)
* inputs (fertilizer, pesticides)

👉 This is now a complete economic layer

# Dataset 6: exchange rates

How exchange rates will be used in this project

This dataset is not for the yield model directly. It is mainly for the economic intelligence layer.

In SQL

It will live as its own year-level table, similar to CPI, fertilizer, and pesticides:

* exchange_rate_table
-------------------
* area
* year
* kes_per_usd

Then it joins on (area, year).

That means when you later build a master table, every crop-year row can inherit the exchange rate for that year.

In the model / analysis layer

Exchange rate matters because your project is partly about market opportunity and supply-demand economics.

It helps in these ways:

1. Interpreting imported crop pressure
If Kenya imports a lot of a crop or input, and the shilling weakens, imported goods become more expensive in local currency. That can:

* raise domestic prices
* create opportunity for local substitutes
* affect the logic around “highly imported crops”

So this dataset strengthens the part of AgriBrain that looks at trade-related market opportunity.

2. Comparing USD and local prices
You already kept producer prices in LCU/tonne, which is good. If later you ever use USD-denominated trade or price datasets, the exchange rate lets you:

* convert USD to KES
* compare external and domestic market signals properly

3. Explaining price changes
Suppose prices rise sharply. Exchange rate helps separate:
* real scarcity from
* currency-driven price increases
* That is similar to CPI, but from the external competitiveness / import-cost angle rather than inflation.

4. LLM reasoning
Your LLM can use this to say things like:
* imported inputs became costlier because the shilling weakened
* crops that rely on imported inputs may face higher production costs
* import-heavy foods may become more expensive, making local alternatives more attractive

So exchange rate is not the “main engine,” but it adds strong economic context.

What this preview tells us
Element looks like only one useful measure: Local currency units per USD
Months includes both Annual value and monthly observations
Unit is NaN, which is fine here because the meaning is already in the element

For this project, we should keep only the annual value, because all your macro tables are being standardized to annual series.

In [ ]:
import pandas as pd

exg_df = pd.read_csv("FAOSTAT_data_exchange rates.csv")
print(exg_df.head())

  Domain Code          Domain  Area Code (M49)   Area ISO Currency Code (FAO)  \
0          PE  Exchange rates              404  Kenya                     KES   
1          PE  Exchange rates              404  Kenya                     KES   
2          PE  Exchange rates              404  Kenya                     KES   
3          PE  Exchange rates              404  Kenya                     KES   
4          PE  Exchange rates              404  Kenya                     KES   

          Currency Element Code                       Element  Year Code  \
0  Kenyan Shilling          LCU  Local currency units per USD       1970   
1  Kenyan Shilling          LCU  Local currency units per USD       1970   
2  Kenyan Shilling          LCU  Local currency units per USD       1970   
3  Kenyan Shilling          LCU  Local currency units per USD       1970   
4  Kenyan Shilling          LCU  Local currency units per USD       1970   

   Year  Months Code        Months  Unit    Value Flag  

In [ ]:
exg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 785 entries, 0 to 784
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Domain Code              785 non-null    object 
 1   Domain                   785 non-null    object 
 2   Area Code (M49)          785 non-null    int64  
 3   Area                     785 non-null    object 
 4   ISO Currency Code (FAO)  785 non-null    object 
 5   Currency                 785 non-null    object 
 6   Element Code             785 non-null    object 
 7   Element                  785 non-null    object 
 8   Year Code                785 non-null    int64  
 9   Year                     785 non-null    int64  
 10  Months Code              785 non-null    int64  
 11  Months                   785 non-null    object 
 12  Unit                     0 non-null      float64
 13  Value                    785 non-null    float64
 14  Flag                     7

In [ ]:
exg_df['Element'].unique()
# exg_df['Item'].unique()

array(['Local currency units per USD',
       'Standard local currency units per USD'], dtype=object)

Step 1: inspect fully

In [ ]:
print("shape:", exg_df.shape)

print("\nunique elements:")
print(exg_df["Element"].value_counts(dropna=False))

print("\nunique months:")
print(exg_df["Months"].value_counts(dropna=False).head(15))

print("\nunique currencies:")
print(exg_df["Currency"].value_counts(dropna=False))

print("\nunique units:")
print(exg_df["Unit"].value_counts(dropna=False))

shape: (785, 16)

unique elements:
Local currency units per USD             729
Standard local currency units per USD     56
Name: Element, dtype: int64

unique months:
Annual value    112
January          57
June             56
December         56
August           56
July             56
February         56
October          56
September        56
November         56
April            56
May              56
March            56
Name: Months, dtype: int64

unique currencies:
Kenyan Shilling    785
Name: Currency, dtype: int64

unique units:
NaN    785
Name: Unit, dtype: int64


🎯 Step 1: Decision

We will KEEP:

👉 Element:

"Local currency units per USD" ✅
(standard one, ignore the “standard local currency…”)

👉 Months:

"Annual value" ✅

Step 2: Filter

In [ ]:
# keep correct element
exg_clean = exg_df[
    exg_df["Element"] == "Local currency units per USD"
]

# keep annual values only
exg_clean = exg_clean[
    exg_clean["Months"] == "Annual value"
]

print("shape:", exg_clean.shape)
print("\ncheck:")
print(exg_clean[["Year", "Value"]].head())

shape: (56, 16)

check:
    Year     Value
0   1970  7.142860
13  1971  7.142860
26  1972  7.142860
39  1973  7.001192
52  1974  7.142860


🎯 What this gives you

Each row:

Kenya
Year
Exchange rate (KES per USD)

👉 Example:

100 → 1 USD = 100 KES
higher value = weaker shilling

Step 3: Keep only needed columns

In [ ]:
exg_clean = exg_clean[[
    "Area",
    "Year",
    "Value"
]]

print(exg_clean.head())
print(exg_clean.columns.tolist())

     Area  Year     Value
0   Kenya  1970  7.142860
13  Kenya  1971  7.142860
26  Kenya  1972  7.142860
39  Kenya  1973  7.001192
52  Kenya  1974  7.142860
['Area', 'Year', 'Value']


What this step does--reduce to:
* area
* year
* exchange rate

👉 removes all metadata noise before final cleaning

Step 4: Final cleaning

In [ ]:
exg_final = exg_clean.copy()

# rename column
exg_final = exg_final.rename(columns={
    "Value": "kes_per_usd"
})

# clean column names
exg_final.columns = [
    col.lower().replace(" ", "_")
    for col in exg_final.columns
]

print(exg_final.head())
print(exg_final.columns.tolist())

     area  year  kes_per_usd
0   Kenya  1970     7.142860
13  Kenya  1971     7.142860
26  Kenya  1972     7.142860
39  Kenya  1973     7.001192
52  Kenya  1974     7.142860
['area', 'year', 'kes_per_usd']


🎯 What you now have

| area | year | kes_per_usd |

👉 Clean, interpretable, and SQL-ready

Step 5: Save

In [ ]:
exg_final.to_csv("exchange_rates_final.csv", index=False)

print("Saved as exchange_rates_final.csv")
print("Final shape:", exg_final.shape)

Saved as exchange_rates_final.csv
Final shape: (56, 3)


Final explanation

This variable:

👉 kes_per_usd

means:

how many Kenyan shillings = 1 USD

So:

higher value → weaker KES
lower value → stronger KES


How it strengthens the system
1. Trade competitiveness

If KES weakens:

imports become expensive
local crops become more competitive

👉 Your model can detect:

“import substitution opportunities”


2. Input costs

Some inputs (fertilizer, pesticides) depend on imports.

👉 Weak KES → higher input costs
👉 affects profitability of crops

3. Price interpretation

If prices rise:

is it real demand?
or currency depreciation?

👉 exchange rate helps answer that

4. LLM reasoning layer

Your system can say:

* “Prices are rising partly due to currency depreciation”
* “Imported crops are becoming expensive, local alternatives are more viable”


Big picture now

We have:

* Food balance → supply/demand
* Producer prices → actual price
* CPI → inflation
* Exchange rate → external price pressure
* Fertilizer → input intensity
* Pesticides → production pressure

👉 This is now a full economic + agronomic + macro system


## Dataset 8: macros

🎯 First: What this dataset is (important framing)

This is broad macro data, not agriculture-specific.

It includes:

GDP
GDP per capita
different price bases (current vs 2015 prices)

👉 So we must be very selective here.

🧠 How macro will be used (clear explanation)

This dataset is mainly for:

1. Context (LLM reasoning)
“economy growing → demand increasing”
“low GDP per capita → affordability constraints”
2. Optional ML features
capturing economic growth trends

👉 BUT:

not core like food balance or prices
do NOT overload your model with too many macro variables

In [ ]:
import pandas as pd

macros_df = pd.read_csv("FAOSTAT_data_macro indicators.csv")
print(macros_df.head())

  Domain Code            Domain  Area Code (M49)   Area  Element Code  \
0          MK  Macro Indicators              404  Kenya          6224   
1          MK  Macro Indicators              404  Kenya          6110   
2          MK  Macro Indicators              404  Kenya          6119   
3          MK  Macro Indicators              404  Kenya          6225   
4          MK  Macro Indicators              404  Kenya          6185   

                                      Element  Item Code  \
0               Value Standard Local Currency      22008   
1                                   Value US$      22008   
2                        Value US$ per capita      22008   
3  Value Standard Local Currency, 2015 prices      22008   
4           Value US$ per capita, 2015 prices      22008   

                     Item  Year Code  Year         Unit         Value Flag  \
0  Gross Domestic Product       1970  1970  million SLC  1.970713e+04    X   
1  Gross Domestic Product       1970  1970  

In [ ]:
macros_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4307 entries, 0 to 4306
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Domain Code       4307 non-null   object 
 1   Domain            4307 non-null   object 
 2   Area Code (M49)   4307 non-null   int64  
 3   Area              4307 non-null   object 
 4   Element Code      4307 non-null   int64  
 5   Element           4307 non-null   object 
 6   Item Code         4307 non-null   int64  
 7   Item              4307 non-null   object 
 8   Year Code         4307 non-null   int64  
 9   Year              4307 non-null   int64  
 10  Unit              4055 non-null   object 
 11  Value             4307 non-null   float64
 12  Flag              4307 non-null   object 
 13  Flag Description  4307 non-null   object 
 14  Note              0 non-null      float64
dtypes: float64(2), int64(5), object(8)
memory usage: 504.9+ KB


In [ ]:
macros_df['Element'].unique()

In [ ]:
macros_df['Item'].unique()

array(['Gross Domestic Product', 'Gross Fixed Capital Formation',
       'Value Added (Agriculture, Forestry and Fishing)',
       'Value Added (Total Manufacturing)',
       'Value Added (Manufacture of food, beverages and tobacco products)',
       'Value Added (Manufacture of food and beverages)',
       'Value Added (Manufacture of tobacco products)',
       'Gross National Income', 'Value Added (Agriculture)',
       'Gross Output (Agriculture, Forestry and Fishing)',
       'Gross Output (Agriculture)'], dtype=object)

Step 1: Inspect fully

In [ ]:
print("shape:", macros_df.shape)

print("\nunique items:")
print(macros_df["Item"].value_counts())

print("\nunique elements:")
print(macros_df["Element"].value_counts())

print("\nunique units:")
print(macros_df["Unit"].value_counts())

shape: (4307, 15)

unique items:
Value Added (Total Manufacturing)                                    752
Value Added (Agriculture, Forestry and Fishing)                      656
Gross Fixed Capital Formation                                        656
Gross Domestic Product                                               654
Value Added (Manufacture of food, beverages and tobacco products)    380
Value Added (Manufacture of food and beverages)                      332
Value Added (Manufacture of tobacco products)                        332
Gross National Income                                                268
Value Added (Agriculture)                                            161
Gross Output (Agriculture)                                            58
Gross Output (Agriculture, Forestry and Fishing)                      58
Name: Item, dtype: int64

unique elements:
Value Standard Local Currency                                                       360
Value US$                        

🎯 What we are looking for

We want to pick ONLY 2–3 strong indicators, typically:

Best candidates:
* GDP (total economy size)
* GDP per capita (income level)
* possibly agriculture value added (if present)


⚠️ What we will drop
* duplicate versions:
* 👉 current vs constant prices (we pick one)
* unnecessary variants:
* multiple currency formats


Likely plan (so you understand ahead)

We will keep:

* GDP (constant prices preferred)
* GDP per capita (constant prices)

👉 These are:

* stable
* comparable across years
* useful signals

### Final selection 

We will keep ONLY 2 variables:

1. GDP (size of economy)

👉 Item:

Gross Domestic Product

👉 Element:

Value US$, 2015 prices



2. Agriculture value added (sector importance)

👉 Item:

Value Added (Agriculture, Forestry and Fishing)

👉 Element:

Value US$, 2015 prices


🧠 Why these two (clear reasoning)
1. GDP
* shows overall demand capacity
* captures macro growth
2. Agriculture value added
* shows strength of agriculture sector
* directly relevant to your project

👉 Together:

give economic context without noise

Step 2: Filter

In [ ]:
# keep only selected items
items_to_keep = [
    "Gross Domestic Product",
    "Value Added (Agriculture, Forestry and Fishing)"
]

macros_clean = macros_df[
    macros_df["Item"].isin(items_to_keep)
]

# keep constant USD values
macros_clean = macros_clean[
    macros_clean["Element"] == "Value US$, 2015 prices"
]

print("shape:", macros_clean.shape)
print("\nitems:")
print(macros_clean["Item"].value_counts())

shape: (110, 15)

items:
Gross Domestic Product                             55
Value Added (Agriculture, Forestry and Fishing)    55
Name: Item, dtype: int64


🎯 What this does

You now isolate:

* clean, comparable macro indicators
* avoid duplicates and inflation issues
* Rows: 55 + 55 = 110 → exactly right

Step 3: Pivot

In [ ]:
macro_pivot = macros_clean.pivot_table(
    index=["Area", "Year"],
    columns="Item",
    values="Value",
    aggfunc="sum"
).reset_index()

# remove column grouping
macro_pivot.columns.name = None

print("shape:", macro_pivot.shape)
print("\ncolumns:")
print(macro_pivot.columns.tolist())

shape: (55, 4)

columns:
['Area', 'Year', 'Gross Domestic Product', 'Value Added (Agriculture, Forestry and Fishing)']


🎯 Expected

Columns like:

* Area
* Year
* Gross Domestic Product
* Value Added (Agriculture, Forestry and Fishing)

Step 4: final cleaning

In [ ]:
macro_final = macro_pivot.copy()

# clean column names
macro_final.columns = [
    col.lower()
       .replace(" ", "_")
       .replace("(", "")
       .replace(")", "")
       .replace(",", "")
    for col in macro_final.columns
]

# rename for readability
macro_final = macro_final.rename(columns={
    "gross_domestic_product": "gdp_usd_2015",
    "value_added_agriculture_forestry_and_fishing": "agriculture_value_added_usd_2015"
})

print(macro_final.head())
print(macro_final.columns.tolist())

    area  year  gdp_usd_2015  agriculture_value_added_usd_2015
0  Kenya  1970  11482.814587                       3409.183768
1  Kenya  1971  12229.307314                       3448.369788
2  Kenya  1972  12844.066030                       3703.078920
3  Kenya  1973  13512.356093                       3868.293211
4  Kenya  1974  14174.783962                       3868.293211
['area', 'year', 'gdp_usd_2015', 'agriculture_value_added_usd_2015']


Step 4: save

In [ ]:
macro_final.to_csv("macro_final.csv", index=False)

print("Saved as macro_final.csv")
print("Final shape:", macro_final.shape)

Saved as macro_final.csv
Final shape: (55, 4)


In SQL, this last table will be used in SQL by joining on (area, year), and in the model mostly as a context layer rather than a core crop signal:

* gdp_usd_2015 helps capture broad demand conditions and economic growth
* agriculture_value_added_usd_2015 helps capture how strong the agriculture sector is in a given year

That matters more for the market/profitability reasoning layer and LLM explanations than for the raw yield model. In practice, this table can help the system distinguish whether a good opportunity is happening within a generally expanding economy or a weaker agricultural period, which adds context to the recommendations.

# SQL

We now have a very solid set of cleaned files: In the brackets are the unique identfiers

1. fdbalance_final.csv (area, item, year)
* Each row = one crop in one year
* This is the main table

2. producer_prices_final.csv (area, item, year)
* Each row = one crop in one year
* Perfect for direct join

3. fertilizer_nutrient_final.csv (area, year)
* One row per year
* Applies to all crops that year

4. pesticides_final.csv (area, year)
* One row per year
* Applies to all crops that year

5. cpi_final.csv (area, year)
* One row per year
* Inflation per year

6. exchange_rates_final.csv (area, year)
* One row per year
* Currency per year

7. macro_final.csv(area, year)
* one row per year
* GDP + agriculture context
* two columns in the same row..the two columns are gdp_usd_2015 and agriculture_value_added_usd_2015

In [ ]:
fdbalance_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1287 entries, 14 to 1342
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   area                      1287 non-null   object 
 1   item                      1287 non-null   object 
 2   year                      1287 non-null   int64  
 3   domestic_supply_quantity  1287 non-null   float64
 4   export_quantity           1216 non-null   float64
 5   feed                      349 non-null    float64
 6   food                      1219 non-null   float64
 7   import_quantity           1203 non-null   float64
 8   losses                    658 non-null    float64
 9   other_uses_non_food       211 non-null    float64
 10  processing                393 non-null    float64
 11  production                1070 non-null   float64
 12  seed                      263 non-null    float64
 13  stock_variation           1094 non-null   float64
 14  supply 

How they will join in SQL (important mental model)

Main table:


* food_balance_table (area, item, year)

Join structure:

* SELECT *
* FROM food_balance_table f

    producer_prices p
* ON f.area = p.area
* AND f.item = p.item
* AND f.year = p.year

* LEFT JOIN fertilizer t
* ON f.area = t.area
* AND f.year = t.year

* LEFT JOIN pesticides ps
* ON f.area = ps.area
* AND f.year = ps.year

* LEFT JOIN cpi c
* ON f.area = c.area
* AND f.year = c.year

* LEFT JOIN exchange_rates e
* ON f.area = e.area
* AND f.year = e.year

* LEFT JOIN macro m
* ON f.area = m.area
* AND f.year = m.year

4. Key insight (very important)

You have two levels of data:

1. Crop-level (high detail)
* food balance
* producer prices

2. Year-level (context)
* fertilizer
* pesticides
* CPI
* exchange rates
* macro

👉 When you join:

year-level data gets broadcast to all crops

## Standardize into lower caps
Before SQL, you can standardize to prevent join issues

## Dataset 7: fertiliser by product

In [ ]:
import pandas as pd

fertprdt_df = pd.read_csv("FAOSTAT_fertiliser by product.csv")
print(fertprdt_df.head())

In [ ]:
fertprdt_df.info()

In [ ]:
fertprdt_df['Element'].unique()

In [ ]:
fertprdt_df['Item'].unique()

In [ ]:
fertprdt_df.columns

## Dataset 9: commodity balances for non food

In [ ]:
import pandas as pd

nonfood_df = pd.read_csv("FAOSTAT_data_commodity balances_non_food.csv")
print(nonfood_df.head())

In [ ]:
nonfood_df.columns


In [ ]:
nonfood_df.info()


In [ ]:
nonfood_df['Element'].unique()


In [ ]:
nonfood_df['Item'].unique()

First impression (from what you showed)

This dataset is:

👉 Commodity Balances (non-food)
👉 Mostly:

fibres (e.g. Abaca)
industrial crops
non-edible materials

This is already a warning sign for your use case

Step 1: Run full inspection (same as before)

In [ ]:
print("shape:", nonfood_df.shape)

print("\nunique units:")
print(nonfood_df["Unit"].value_counts(dropna=False))

print("\nunique elements:")
print(nonfood_df["Element"].value_counts(dropna=False))

print("\nsample items:")
print(nonfood_df["Item"].unique()[:30])

shape: (6360, 14)

unique units:
t    6360
Name: Unit, dtype: int64

unique elements:
Domestic supply quantity         1272
Import quantity                  1219
Export quantity                  1166
Other uses (non-food)             689
Stock Variation                   689
Production                        636
Feed                              477
Food supply quantity (tonnes)      53
Losses                             53
Seed                               53
Processing                         53
Name: Element, dtype: int64

sample items:
['Abaca' 'Alcohol, Non-Food' 'Brans' 'Copra Cake' 'Cotton lint'
 'Cottonseed' 'Cottonseed Cake' 'Groundnut Cake' 'Hard Fibres, Other'
 'Jute' 'Oilseed Cakes, Other' 'Palm kernels' 'Palmkernel Cake'
 'Rape and Mustard Cake' 'Rape and Mustardseed' 'Rubber' 'Sesameseed Cake'
 'Silk' 'Sisal' 'Soft-Fibres, Other' 'Soyabean Cake' 'Sunflowerseed Cake'
 'Tobacco' 'Wool (Clean Eq.)']


Final verdict on this dataset

👉 We do NOT use this dataset (as a whole)

Why (very clear reasoning)

Look at your items:

Abaca
Cotton lint
Jute
Sisal
Rubber
Silk
Oilseed cakes
Wool

👉 These are:

industrial fibres
processing by-products
non-edible commodities
Why this is a problem for AgriBrain

Your goal is:

recommend crops farmers can plant for profit

This dataset mostly represents:

post-processing outputs (cakes, fibres)
industrial chains, not farm decisions

Example:

“Cotton lint” → processed output
Farmer grows → cottonseed, not lint
The key issue

If you use this dataset:

your model may recommend non-plantable items
your LLM reasoning becomes confused
your economic signals become noisy
Are there ANY useful items?

A few borderline ones:

Cottonseed
Rape and Mustardseed
Sesameseed
Tobacco

👉 But:

these will already appear in your food balance dataset
or can be handled later if needed

In [ ]:
# Non-food FAOSTAT dataset excluded due to:
# - mostly industrial/non-plantable items
# - not aligned with farmer recommendation objective

In [15]:
df=pd.read_csv('Test\kamis_pulse_20260408.csv')
df.head(100)
#df.columns


,Commodity,Classification,Grade,Sex,Market,Wholesale,Retail,Supply Volume,County,Date,mapped_commodity
0,Dry Maize,White Maize,-,-,Embu Town,50.00,65.00,7000.0,Embu,2026-04-08,Maize (white)
1,Dry Maize,White Maize,-,-,Ahero,65.00,70.00,45000.0,Kisumu,2026-04-07,Maize (white)
2,Dry Maize,Mixed-Traditional,-,-,Ahero,65.00,70.00,5000.0,Kisumu,2026-04-07,Maize (white)
3,Dry Maize,White Maize,-,-,Kathonzweni,50.00,55.00,9000.0,Makueni,2026-04-07,Maize (white)
4,Dry Maize,White Maize,-,-,Mogogosiek Market,41.67,58.33,10800.0,Bomet,2026-04-07,Maize (white)
...,...,...,...,...,...,...,...,...,...,...,...
95,Dry Maize,White Maize,-,-,Mogogosiek Market,41.67,54.17,12000.0,Bomet,2026-03-17,Maize (white)
96,Dry Maize,White Maize,-,-,Kerugoya,48.00,75.00,4400.0,Kirinyaga,2026-03-17,Maize (white)
97,Dry Maize,White Maize,-,-,Ahero,65.00,70.00,55000.0,Kisumu,2026-03-17,Maize (white)
98,Dry Maize,Mixed-Traditional,-,-,Ahero,65.00,70.00,7000.0,Kisumu,2026-03-17,Maize (white)


In [16]:
df['Commodity'].unique()

<StringArray>
[           'Dry Maize',                 'Rice',    'Kales/Sukuma Wiki',
 'White Irish Potatoes',          'Maize Flour',          'Wheat Flour']
Length: 6, dtype: str

In [17]:
df.shape

(30000, 11)

In [18]:
df

,Commodity,Classification,Grade,Sex,Market,Wholesale,Retail,Supply Volume,County,Date,mapped_commodity
0,Dry Maize,White Maize,-,-,Embu Town,50.00,65.00,7000.0,Embu,2026-04-08,Maize (white)
1,Dry Maize,White Maize,-,-,Ahero,65.00,70.00,45000.0,Kisumu,2026-04-07,Maize (white)
2,Dry Maize,Mixed-Traditional,-,-,Ahero,65.00,70.00,5000.0,Kisumu,2026-04-07,Maize (white)
3,Dry Maize,White Maize,-,-,Kathonzweni,50.00,55.00,9000.0,Makueni,2026-04-07,Maize (white)
4,Dry Maize,White Maize,-,-,Mogogosiek Market,41.67,58.33,10800.0,Bomet,2026-04-07,Maize (white)
...,...,...,...,...,...,...,...,...,...,...,...
29995,Wheat Flour,Exe brown bread,-,-,Nairobi Supermarkets,0.00,100.00,NaN,Nairobi,2024-06-21,Wheat flour
29996,Wheat Flour,Exe Atta Mark 1,-,-,Nairobi Supermarkets,0.00,99.50,NaN,Nairobi,2024-06-21,Wheat flour
29997,Wheat Flour,Exe All purpose fortified wheat flour,-,-,Nairobi Supermarkets,0.00,94.50,NaN,Nairobi,2024-06-21,Wheat flour
29998,Wheat Flour,Elliots self raising flour,-,-,Nairobi Supermarkets,0.00,92.50,NaN,Nairobi,2024-06-21,Wheat flour
